File: qail-bench/benchmarks/pool_benchmark.c
/*
 * C libpq POOL BENCHMARK
 * 
 * libpq doesn't have built-in pooling, so we simulate with:
 * - Pre-created connections (1 per thread)
 * - Pipelining within each connection
 * 
 * Compile: gcc -O3 -march=native -o pool_benchmark pool_benchmark.c -lpq -lpthread
 * Run: ./pool_benchmark
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libpq-fe.h>
#include <pthread.h>
#include <time.h>
#include <stdatomic.h>

#define TOTAL_QUERIES 150000000
#define NUM_WORKERS 10
#define POOL_SIZE 10
#define QUERIES_PER_BATCH 100

static atomic_long counter = 0;
static int batches_per_worker;

typedef struct {
    int worker_id;
    const char *conninfo;
} WorkerArgs;

void* worker_thread(void* arg) {
    WorkerArgs* args = (WorkerArgs*)arg;
    
    // Each worker has its own connection (simulating pool)
    PGconn* conn = PQconnectdb(args->conninfo);
    if (PQstatus(conn) != CONNECTION_OK) {
        fprintf(stderr, "Worker %d: Connection failed: %s\n", args->worker_id, PQerrorMessage(conn));
        PQfinish(conn);
        return NULL;
    }
    
    // Prepare statement BEFORE pipeline mode
    const char* stmt_name = "select_harbors";
    const char* sql = "SELECT id, name FROM harbors LIMIT $1";
    PGresult* res = PQprepare(conn, stmt_name, sql, 1, NULL);
    if (PQresultStatus(res) != PGRES_COMMAND_OK) {
        fprintf(stderr, "Worker %d: Prepare failed: %s\n", args->worker_id, PQerrorMessage(conn));
        PQclear(res);
        PQfinish(conn);
        return NULL;
    }
    PQclear(res);
    
    // Enable pipelining AFTER prepare
    if (PQenterPipelineMode(conn) != 1) {
        fprintf(stderr, "Worker %d: Failed to enter pipeline mode\n", args->worker_id);
        PQfinish(conn);
        return NULL;
    }
    
    // Execute batches
    for (int b = 0; b < batches_per_worker; b++) {
        // Send batch of queries
        for (int i = 0; i < QUERIES_PER_BATCH; i++) {
            char limit_str[16];
            snprintf(limit_str, sizeof(limit_str), "%d", (i % 10) + 1);
            const char* params[1] = { limit_str };
            
            if (PQsendQueryPrepared(conn, stmt_name, 1, params, NULL, NULL, 0) != 1) {
                fprintf(stderr, "Worker %d: Send failed\n", args->worker_id);
                break;
            }
        }
        
        // Sync pipeline and flush
        PQpipelineSync(conn);
        PQflush(conn);
        
        // Consume results
        for (int q = 0; q < QUERIES_PER_BATCH; q++) {
            // Each query: get result until NULL
            PGresult* result;
            while ((result = PQgetResult(conn)) != NULL) {
                ExecStatusType status = PQresultStatus(result);
                if (status == PGRES_TUPLES_OK) {
                    // Consume rows
                    int nrows = PQntuples(result);
                    for (int r = 0; r < nrows; r++) {
                        char* id = PQgetvalue(result, r, 0);
                        char* name = PQgetvalue(result, r, 1);
                        (void)id; (void)name;
                    }
                    atomic_fetch_add(&counter, 1);
                }
                PQclear(result);
            }
        }
        
        // Consume pipeline sync
        PGresult* sync_result = PQgetResult(conn);
        if (sync_result) {
            PQclear(sync_result);
        }
    }
    
    PQexitPipelineMode(conn);
    PQfinish(conn);
    return NULL;
}

int main() {
    const char* host = getenv("PG_HOST") ? getenv("PG_HOST") : "127.0.0.1";
    const char* port = getenv("PG_PORT") ? getenv("PG_PORT") : "5432";
    const char* user = getenv("PG_USER") ? getenv("PG_USER") : "postgres";
    const char* database = getenv("PG_DATABASE") ? getenv("PG_DATABASE") : "postgres";
    
    char conninfo[512];
    snprintf(conninfo, sizeof(conninfo), "host=%s port=%s user=%s dbname=%s", 
             host, port, user, database);
    
    printf("🔌 Connecting to %s:%s as %s\n", host, port, user);
    printf("🚀 C LIBPQ POOL BENCHMARK\n");
    printf("=========================\n");
    printf("Total queries:    %15d\n", TOTAL_QUERIES);
    printf("Workers:          %15d\n", NUM_WORKERS);
    printf("Pool size:        %15d\n", POOL_SIZE);
    printf("Batch size:       %15d\n", QUERIES_PER_BATCH);
    printf("\n");
    
    batches_per_worker = TOTAL_QUERIES / NUM_WORKERS / QUERIES_PER_BATCH;
    
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    
    // Create worker threads
    pthread_t threads[NUM_WORKERS];
    WorkerArgs args[NUM_WORKERS];
    
    for (int i = 0; i < NUM_WORKERS; i++) {
        args[i].worker_id = i;
        args[i].conninfo = conninfo;
        pthread_create(&threads[i], NULL, worker_thread, &args[i]);
    }
    
    // Wait for all threads
    for (int i = 0; i < NUM_WORKERS; i++) {
        pthread_join(threads[i], NULL);
    }
    
    clock_gettime(CLOCK_MONOTONIC, &end);
    
    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    double qps = counter / elapsed;
    
    printf("\n📈 FINAL RESULTS:\n");
    printf("┌──────────────────────────────────────────────────┐\n");
    printf("│ C LIBPQ POOL BENCHMARK                           │\n");
    printf("├──────────────────────────────────────────────────┤\n");
    printf("│ Total Time:               %15.1fs │\n", elapsed);
    printf("│ Queries/Second:           %15.0f │\n", qps);
    printf("│ Workers:                  %15d │\n", NUM_WORKERS);
    printf("│ Pool Size:                %15d │\n", POOL_SIZE);
    printf("│ Queries Completed:        %15ld │\n", counter);
    printf("└──────────────────────────────────────────────────┘\n");
    
    return 0;
}

File: qail-bench/benchmarks/pool_benchmark.go
// Go pgx POOL BENCHMARK (BATCHED)
//
// Tests connection pool performance with concurrent workers.
// Uses pgx.Batch to be fair with QAIL pipelining.
//
// Run: go run pool_benchmark.go

package main

import (
	"context"
	"fmt"
	"os"
	"sync"
	"sync/atomic"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

const (
	TOTAL_QUERIES     = 150_000_000
	NUM_WORKERS       = 10
	POOL_SIZE         = 10
	QUERIES_PER_BATCH = 100
)

func getEnvOr(key, defaultVal string) string {
	if val := os.Getenv(key); val != "" {
		return val
	}
	return defaultVal
}

func main() {
	ctx := context.Background()

	host := getEnvOr("PG_HOST", "127.0.0.1")
	port := getEnvOr("PG_PORT", "5432")
	user := getEnvOr("PG_USER", "postgres")
	database := getEnvOr("PG_DATABASE", "postgres")

	connStr := fmt.Sprintf("postgres://%s@%s:%s/%s?pool_max_conns=%d&pool_min_conns=%d",
		user, host, port, database, POOL_SIZE, POOL_SIZE)

	fmt.Printf("🔌 Connecting to %s:%s as %s\n", host, port, user)

	pool, err := pgxpool.New(ctx, connStr)
	if err != nil {
		panic(err)
	}
	defer pool.Close()

	fmt.Println("🚀 GO PGX POOL BENCHMARK (BATCHED)")
	fmt.Println("==================================")
	fmt.Printf("Total queries:    %15d\n", TOTAL_QUERIES)
	fmt.Printf("Workers:          %15d\n", NUM_WORKERS)
	fmt.Printf("Pool size:        %15d\n", POOL_SIZE)
	fmt.Printf("Batch size:       %15d\n", QUERIES_PER_BATCH)
	fmt.Println()

	batchesPerWorker := TOTAL_QUERIES / NUM_WORKERS / QUERIES_PER_BATCH
	var counter int64
	var wg sync.WaitGroup

	start := time.Now()

	// Spawn workers
	for w := 0; w < NUM_WORKERS; w++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			
			for b := 0; b < batchesPerWorker; b++ {
				// Acquire connection
				conn, err := pool.Acquire(ctx)
				if err != nil {
					fmt.Printf("Worker %d: acquire failed: %v\n", workerID, err)
					return
				}

				// Build Batch
				batch := &pgx.Batch{}
				for i := 0; i < QUERIES_PER_BATCH; i++ {
					limit := (i % 10) + 1
					batch.Queue("SELECT id, name FROM harbors LIMIT $1", limit)
				}

				// Send Batch
				br := conn.SendBatch(ctx, batch)
				
				// Consume Results
				for i := 0; i < QUERIES_PER_BATCH; i++ {
					rows, err := br.Query()
					if err != nil {
						fmt.Printf("Worker %d: batch query failed: %v\n", workerID, err)
						break
					}
					// Consume rows
					for rows.Next() {
						var id string
						var name string
						rows.Scan(&id, &name)
					}
					rows.Close()
				}
				
				br.Close()
				conn.Release()
				atomic.AddInt64(&counter, int64(QUERIES_PER_BATCH))
			}
		}(w)
	}

	// Progress reporter
	done := make(chan bool)
	go func() {
		for {
			select {
			case <-done:
				return
			case <-time.After(2 * time.Second):
				count := atomic.LoadInt64(&counter)
				if count >= TOTAL_QUERIES {
					return
				}
				elapsed := time.Since(start).Seconds()
				qps := float64(count) / elapsed
				remaining := TOTAL_QUERIES - int(count)
				eta := float64(remaining) / qps
				fmt.Printf("   %6d queries | %8.0f q/s | ETA: %.0fs\n", count, qps, eta)
			}
		}
	}()

	wg.Wait()
	done <- true

	elapsed := time.Since(start)
	qps := float64(TOTAL_QUERIES) / elapsed.Seconds()

	fmt.Println("\n📈 FINAL RESULTS:")
	fmt.Println("┌──────────────────────────────────────────────────┐")
	fmt.Println("│ GO PGX POOL BENCHMARK (BATCHED)                  │")
	fmt.Println("├──────────────────────────────────────────────────┤")
	fmt.Printf("│ Total Time:               %15.1fs │\n", elapsed.Seconds())
	fmt.Printf("│ Queries/Second:           %15.0f │\n", qps)
	fmt.Printf("│ Workers:                  %15d │\n", NUM_WORKERS)
	fmt.Printf("│ Pool Size:                %15d │\n", POOL_SIZE)
	fmt.Printf("│ Queries Completed:        %15d │\n", atomic.LoadInt64(&counter))
	fmt.Println("└──────────────────────────────────────────────────┘")
}
